crates/lib3h/src/engine/real_engine.rs

```diff
@@ -0,0 +1,420 @@
+            Lib3hClientProtocol::HandleGetGossipingEntryListResult(_msg) => {
+                // FIXME
+            }
+        }
+        Ok((did_work, outbox))
+    }
+
+    /// Create a gateway for this agent in this space, if not already part of it.
+    /// Must not already be part of this space.
+    fn serve_JoinSpace(&mut self, join_msg: &SpaceData) -> Lib3hResult<Lib3hServerProtocol> {
+        // Prepare response
+        let mut res = GenericResultData {
+            request_id: join_msg.request_id.clone(),
+            space_address: join_msg.space_address.clone(),
+            to_agent_id: join_msg.agent_id.clone(),
+            result_info: vec![],
+        };
+        // Bail if space already joined by agent
+        let chain_id = (join_msg.space_address.clone(), join_msg.agent_id.clone());
+        if self.space_gateway_map.contains_key(&chain_id) {
+            res.result_info = "Already joined space".to_string().into_bytes();
+            return Ok(Lib3hServerProtocol::FailureResult(res));
+        }
+        // First create DhtConfig for space gateway
+        let agent_id = std::string::String::from_utf8_lossy(&join_msg.agent_id).into_owned();
+        let this_net_peer = self.network_gateway.borrow().this_peer().clone();
+        let dht_config = DhtConfig {
+            this_peer_address: agent_id,
+            this_peer_transport: this_net_peer.peer_address.clone(),
+            custom: self.config.dht_custom_config.clone(),
+        };
+        // Create new space gateway for this ChainId
+        let new_space_gateway = P2pGateway::new_with_space(
+            Rc::clone(&self.network_gateway),
+            &join_msg.space_address,
+            self.dht_factory,
+            &dht_config,
+        );
+
+        // HACK: Send JoinSpace to all known peers
+        let space_address =
+            std::string::String::from_utf8_lossy(&join_msg.space_address).into_owned();
+        let peer = new_space_gateway.this_peer().to_owned();
+        let mut payload = Vec::new();
+        let p2p_msg = P2pProtocol::JoinSpace(space_address.clone(), peer.clone());
+        p2p_msg
+            .serialize(&mut Serializer::new(&mut payload))
+            .unwrap();
+        println!(
+            "[t] {} - Broadcasting JoinSpace: {}, {}",
+            self.name.clone(),
+            space_address,
+            peer.peer_address
+        );
+        self.network_gateway.borrow_mut().send_all(&payload).ok();
+        // HACK END
+
+        // Add it to space map
+        self.space_gateway_map
+            .insert(chain_id.clone(), new_space_gateway);
+        // Have DHT broadcast our PeerData
+        let space_gateway = self.space_gateway_map.get_mut(&chain_id).unwrap();
+        Dht::post(
+            space_gateway,
+            DhtCommand::HoldPeer(PeerData {
+                peer_address: dht_config.this_peer_address,
+                transport: dht_config.this_peer_transport,
+                timestamp: 42, // FIXME
```

Comment:
rust's time functions are so clumsy...
```rust
  let ms = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap();
  let ms = ms.as_secs() as u64 * 1000 + ms.subsec_nanos() as u64 / 1_000_000;
```
it takes 2 variable binds to get milliseconds since the epoch...

Reply:
Should we have our own wrapper for timestamping?